### ISSUE TITLE
Adaptive convection resulting in extremely low zone temperature which causes fatal error

### ISSUE THREAD

**Issue by mjwitte** [NREL/EnergyPlus#5906]
## Issue overview

This user file fails due to a very cold zone air temperature.

```
   ** Warning ** Temperature out of range [-100. to 200.] (PsyPsatFnTemp)
   **   ~~~   **  Routine=CorrectZoneHumRat, Environment=RUN PERIOD 1, at Simulation time=01/12 00:35 - 00:40
   **   ~~~   **  Input Temperature=-263.23
   ** Warning ** Calculated Relative Humidity out of range (PsyRhFnTdbWPb) 
   **   ~~~   **  Routine=CorrectZoneAirTemp, Environment=RUN PERIOD 1, at Simulation time=01/12 00:35 - 00:40
   **   ~~~   **  Dry-Bulb= -263.23 Humidity Ratio= 1.000E-005 Calculated Relative Humidity [%]= 93059.94
   **   ~~~   ** Relative Humidity being reset to 100.0%
   ** Severe  ** PsyRhoAirFnPbTdbW: RhoAir (Density of Air) is calculated <= 0 [-6.29394].
   **   ~~~   ** pb =[99033.33], tdb=[-327.74], w=[2.6406651E-003].
   **   ~~~   **  Routine=CorrectZoneHumRat, Environment=RUN PERIOD 1, at Simulation time=01/16 23:35 - 23:40
   **  Fatal  ** Program terminates due to preceding condition.
```

Running this in the debugger shows that the error is in  in zone 10 (ZoneWest) and the cold temperature is calculated in ZoneTempPredictorCorrector::CorrectZoneAirTemp

```
                TempDepCoef = SumHA + SumMCp + SumSysMCp;
                TempIndCoef = SumIntGain + SumHATsurf - SumHATref + SumMCpT + SumSysMCpT + ( NonAirSystemResponse( ZoneNum ) / ZoneMult + SysDepZoneLoadsLagged( ZoneNum ) );
```

At the time of the error:
TempDepCoef = 0
TempIndCoef = -395
SumHATsurf = 6481
SumHATref = 6876
and all the other terms above are zero.

Turning off interior convection = AdaptiveConvectionAlgorithm gets rid of the error and the file runs to completion, so it _appears_ that the problem is in adaptive convection.
### Details

Some additional details for this issue (if relevant):
- Platform Win64
- Version of EnergyPlus 8.6.0
- helpdesk ticket number 11490
### Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [ ] Defect file added to EnergyPlusDevSupport\DefectFiles  
  _Note that the defect file will fail in the debugger due to #5905 - so either need to add ChW connections to the first radiant system, or change LowTempRadiantSystem.cc line 5102 to `} else if ( ( SELECT_CASE_var1 == CoolingMode ) && HydrRadSys(RadSysNum).CoolingSystem ) {` so it gets past that problem if #5905 isn't fixed before this._
- [ ] Ticket added to Pivotal for defect (development team task)
- [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

**Comment by mjwitte** [comment]
_Additional information from the user:_

the following observation might help to identify the deeper root(s) of
the problem:

If the simulation is still run with 

SurfaceConvectionAlgorithm:Inside=AdaptiveConvectionAlgorithm,

but with the ZoneCrossMixing Shutoff-options deactivated, i.e.

DeltaTemperature=0
optional (shutoff condition) schedules deleted,

then the simulation also runs without fatal error.
I tested this with location=DENVER.

**Comment by lgu1234** [comment]
@mjwitte After preliminary testing, I found that the cause is to incorrectly calculate surface convection load when Surface( SurfNum ).TAirRef == ZoneSupplyAirTemp. Your observation shows this:

SumHA = 0 in TempDepCoef
SumHATsurf = hc*A*Tsurf
SumHARef = hc*A*RefAirTemp

SumHATsurf - SumHARef (used in TempIndCoef)
 
I think it has two mistakes. The first mistake is no SumHA calculation. SumHA should be applied to every surface. The second mistake is incorrect calculation of SumHARef. It should be temperature difference.  

Surface convection load should be calculated in the following items in the zone energy balance:

Sum hc*A*(Tzone - Tsurf)
If Surface( SurfNum ).TAirRef == ZoneSupplyAirTemp, Tsurf should be replaced by Tref. 

The code modification is 

			SumHATsurf += HConvIn( SurfNum ) * Area * TempSurfInTmp( SurfNum );
....

 if ( SELECT_CASE_var == ZoneSupplyAirTemp ) {
			SumHA += HA;
			SumHATref += HA * ( TempSurfInTmp( SurfNum ) - RefAirTemp );
}

The testing results show normal zone temperature and no fatal error. The program ran to completion.

Please let me know what you think.

**Comment by lgu1234** [comment]
@mjwitte @EnergyArchmage I though of this issue and did some testing. Here is what I think.

After adding an output variable of "Surface Inside Face Convection Heat Transfer Coefficient", I ran the test again to see what values of film coefficients are in the problem zone: ZoneWest. The zone has six surfaces. The following figure show min and max values of inside surface convective coefficients in all 6 surface before the program crash:
![convectivecoeff](https://cloud.githubusercontent.com/assets/5970656/22991626/ee613410-f38a-11e6-84b3-8a366bb71965.JPG)
The values are normal, even though the largest value is 43. Therefore, the problem is not from adaptive convection.

Then I think of zone energy balance equation based on this particular condition: rho*Cp*dT/dt (Zone air Thermal storage) = SumHATsurf - SumHATref. Since both items are not related to zone temperature, the difference of both items is treated as internal gain. If the difference is negative, zone air temperature at the current time step (correct procedure) is always below one at previous time step. And vise versa. In this case in winter time, RefAirTemp (from inlet node temperature) is always higher than the inside surface temperature, that is why the zone temperature is dropped below -100C. Therefore, the existing calculation approach may not be correct. Here are possible solutions:

1. Treat the case of ZoneSupplyAirTemp as  ZoneMeanAirTemp, so that SumHATsurf - SumHATref becomes SumHATsurf - SumHA * Tzone
2. Assume supply air temperature replaces inside surface temperature as reference, so that convection heat transfer becomes: SumHATref - SumHA * Tzone
3. Ignore this item in zone energy balance, add the convection gain as heat flux in Surface Heat Balance.

Personally, I prefer item 2. I ran the test file with my modification if Item 2 to completion. No strange zone temperature results are found.

Please let me know your thought. Although my suggestion is not perfect, it is better than the current approach.

**Comment by mjwitte** [comment]
@lgu1234 Some intial comments about the proposed fix.  Sorry this is long - documenting the train of thought here.

1. I think the original code is correct to *not* increment SumHA for surfaces that use RefAirTemp (i.e. the convection is not dependent on zone air temperature), because SumHA is part of TempDepCoef which is dependent on zone temperature:
`TempDepCoef = SumHA + SumMCp`,

2. I do not think a deltaT belongs in SumHATRef, because it is subtracted from SumHATSurf here
`TempIndCoef = SumIntGain + SumHATsurf - SumHATref + SumMCpT + SumSysMCpT + ( NonAirSystemResponse( ZoneNum ) / ZoneMult + SysDepZoneLoadsLagged( ZoneNum ) );`

3. Looking  in [CorrectZoneAirTemp](https://github.com/NREL/EnergyPlus/blob/88cd2033f8f4fb2871e4f36a68e2fee0d606dcb6/src/EnergyPlus/ZoneTempPredictorCorrector.cc#L4310-L4315) at the simplest case of UseAnalyticalSolution (which the defect file is using) and TempDepCoef == 0.0, then the equation for ZT is simply
`ZT( ZoneNum ) = ZoneT1( ZoneNum ) + TempIndCoef / AirCap`

If you extract `SumHATSurf - SumHATref` as a single term from the above equation, then SumHATSurf should only include surfaces that use Tref, not all surfaces.

4. If I follow current code, and all of the surfaces in the zone use ZoneSupplyAirTemp as Tref, then look at the case where there is no airflow from the system or infiltration or mixing.

RefAirTemp gets set to the last inlet node temp [here](https://github.com/NREL/EnergyPlus/blob/88cd2033f8f4fb2871e4f36a68e2fee0d606dcb6/src/EnergyPlus/ZoneTempPredictorCorrector.cc#L5551-L5556),  With the original code SumHA is zero (because it's skipped for these surfaces), SumMCp is zero (because there is no other non-system airflow), so TempDepCoef=0.

If nothing else is happening in the zone (no internal gains, etc.), then 
`TempIndCoef = SumHATsurf - SumHATref`
which means the node temperature will drive changes in the zone air temperature even though there is no system flow.

5.  Digging a bit deeper, when adaptive convection selects one of the FisherPedersenCeilDiffuser models, it should be setting the surface reference temp to the zone air temperature, *not* the supply temp, per the engineering ref for the Ceiling Diffuser Algorithm "The correlation was reformulated to use the room outlet temperature as
the reference temperature."  If I'm following correctly, all [three of these if blocks](https://github.com/NREL/EnergyPlus/blob/ded3c56a08801ac56fca2da93ac92e775686acae/src/EnergyPlus/ConvectionCoefficients.cc#L5603-L5651) should always set `Surface( SurfNum ).TAirRef = ZoneMeanAirTemp`. @EnergyArchmage ?

6. In ConvectionCoefficients.hh, is this comment still applicable? @RKStrand ?
```
	// CalcCeilingDiffuserInletCorr should replace CalcCeilingDiffuser (above), if ZoneTempPredictorCorrector can
	// ever be made to work correctly with the inlet air temperature.
```
If so, this calls into question the way that other correlations which *do* use ZoneSupplyAirTemp as the reference temperature, which would be HcInt_GoldsteinNovoselacCeilingDiffuserWindow/Walls/Floow and potentially user defined convection.

**Comment by lgu1234** [comment]
@mjwitte Here are my thought related to your comments.

Everything what you mention is correct in the code as designed. Let me use a simplest equation to show you what I think: 

ZT( ZoneNum ) = ZoneT1( ZoneNum ) + TempIndCoef / AirCap
TempIndCoef = SumHATsurf - SumHATref

Let me assume a heating case, the supply air temperature is greater than a surface temperature and greater than zone air temperature. Physically, it provides heating. Then Tref > Tsurf and Tref > ZoneT1( ZoneNum ). Pure heating. The value of TempIndCoef (Tsurf - Tref) is negative. When the zone equation is solved, ZT( ZoneNum )  < ZoneT1( ZoneNum ) under heating condition, due to the negative value of TempIndCoef . That was why the issue shows very low zone air temperature. 

I don't see any code wrong. What I want to say is that the concept of physics is wrong. What I propose is not a perfect solution, but a much better solution than the existing one.

**Comment by mjwitte** [comment]
@lgu1234 But if the supply air temperature is warmer than the surface temperature, that will cause the convective heat flow at the surface to be away from the zone air mass and to the surface.  So, it *is* removing heat from the zone air mass.  I think the equation is OK as-is.  

So, I go back to this code (which appears in multiple places):
```
			// a weighted average of the inlet temperatures.
			if ( SumSysMCp > 0.0 ) { // protect div by zero
				RefAirTemp( SurfNum ) = SumSysMCpT / SumSysMCp; // BG changed 02-16-2005 to add index (SurfNum)
			} else {
				RefAirTemp( SurfNum ) = NodeTemp;
			}
```
If there is no system airflow, then the NodeTemp is irrelevant.  Seems more appropriate to set this to MAT.

**Comment by lgu1234** [comment]
@mjwitte I think this is a solution, so that the code may become:

			} else if ( SELECT_CASE_var == ZoneSupplyAirTemp ) {
				// check whether this zone is a controlled zone or not
				if ( ! ControlledZoneAirFlag ) {
					ShowFatalError( "Zones must be controlled for Ceiling-Diffuser Convection model. No system serves zone " + Zone( ZoneNum ).Name );
					return;
				}
				// determine supply air temperature as a weighted average of the inlet temperatures.
				if ( SumSysMCp > 0.0 ) {
					RefAirTemp = SumSysMCpT / SumSysMCp;
					SumHATref += HA * RefAirTemp;
				} else {
					// no system flow (yet) so just use last value for inlet node temp, this can happen early in the environment
					// #5906
					SumHA += HA;
				}

I need to run the test case to see if the file still crashes or not.

**Comment by mjwitte** [comment]
@lgu1234 Yes, that looks like a better solution.  But there may still need to be changes to other similar code when SumSysMCp is zero?

**Comment by lgu1234** [comment]
@mjwitte I ran the test file to completion without crash. The zone air temperatures of the interest vary between 14 and 61C, similar to other zones. It is a better solution. 

I also check cases with SumSysMCp and find following modules use 			if ( SumSysMCp > 0.0 ) { // protect div by zero
HeatBalanceSurfaceManager
WindowManager
RoomAirModelAirflowNetwork
and 
CalcZoneComponentLoadSums in ZoneTempPredictorCorrector.

I am going to fix CalcZoneComponentLoadSums in ZoneTempPredictorCorrector for this issue and fix others with new issues.

**Comment by mjwitte** [comment]
Closed via #6026
